Update s3 secure settings

albertzaharovits opened the pull request:
Secure settings for the S3 repository plugin are update-able.
AWSS3 clients have to be requested from the `InternalAwsS3Service` service for each use. These client instances cannot be stored and used at a later time because they can change if their settings change. The client request and client update logic is protected by a read-write lock.

TODOs:
- [x] fix existing tests
- [x] add tests for reload credentials
- [x] ~fail if removing settings in use by the repository ?~
- [x] ~keep existing settings override from cluster settings ?~

s1monw commented on plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/InternalAwsS3Service.java:
```diff
@@ -28,66 +28,93 @@
-
-    private final Map<String, S3ClientSettings> clientsSettings;
+    private final ReadWriteLock lock;
+    private final ConcurrentHashMap<String, AmazonS3> clientsCache;
```
lets make this an immutable map instead. That way we don't need a read-write lock and can just make the method that changes the reference synchronized. Also for the `AmazonS3Wrapper` we can just use a reference counter. We already have the infrastructure for it in `org.elasticsearch.common.util.concurrent.AbstractRefCounted` lemme know if you need help with this.

s1monw commented on plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/InternalAwsS3Service.java:
```diff
@@ -28,66 +28,93 @@
+    public void updateClientSettings(Settings settings) {
+        lock.writeLock().lock();
+        try {
+            // clear previously cached clients
```
++ I like that we completely destroy it once we are refreshing the settings.

s1monw commented on plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java:
```diff
@@ -116,63 +114,63 @@ public void deleteBlob(String blobName) throws IOException {
 
     @Override
     public Map<String, BlobMetaData> listBlobsByPrefix(@Nullable String blobNamePrefix) throws IOException {
-        return AccessController.doPrivileged((PrivilegedAction<Map<String, BlobMetaData>>) () -> {
```
can we please not change the way we use AccessController here. We can do this in a sep. change but this is a cause of trouble and hard to understand why it's needed if at all

albertzaharovits commented:
@s1monw I have addressed your feedback.
The client cache is a volatile reference to an immutable map.
Also, `AmazonS3Wrapper`interface extends `RefCounted` *and* `Releasable`. The implementation extends `AbstractRefCount`. The `Releasable#close` methods is a proxy for `RefCounted#decRef` so it's not really required ,but it's nice because it allows the usage inside try/with blocks.

albertzaharovits commented:
@s1monw this is ready for another review round. Can you please take a look?
If you agree if the approach I will update the tests as well.

s1monw commented on plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/InternalAwsS3Service.java:
```diff
@@ -28,66 +28,78 @@
-            return client;
+    public synchronized void updateClientSettings(Settings settings) {
+        // the clients will shutdown when they will not be used anymore
+        for (final AmazonS3Wrapper clientWrapper : clientsCache.values()) {
```
can u use Releasables#close here instead of a loop.

s1monw commented on plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/InternalAwsS3Service.java:
```diff
@@ -28,66 +28,78 @@
+        // clear previously cached clients
+        clientsCache = Collections.unmodifiableMap(emptyMap());
+        // reload secure settings
+        clientsSettings = Collections.unmodifiableMap(S3ClientSettings.load(settings));
```
leave a comment here that we built the clients lazily

s1monw commented on plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/InternalAwsS3Service.java:
```diff
@@ -28,66 +28,78 @@
+            clientWrapper.decRef();
         }
+        // clear previously cached clients
+        clientsCache = Collections.unmodifiableMap(emptyMap());
```
Collections.emptyMap() is already immutable

s1monw commented on plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/InternalAwsS3Service.java:
```diff
@@ -174,4 +166,30 @@ public void refresh() {
         }
     }
+
+    private static class InternalAmazonS3Wrapper extends AbstractRefCounted implements AmazonS3Wrapper {
```
I don't think we need the `AmazonS3Wrapper` interface just make this class public and name it `AmazonS3Wrapper`

s1monw replied:
I'd call it `AmazonS3Reference`

albertzaharovits commented:
test this please
